[dtensor] fix new_empty_strided op

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #107835

This PR fixes the new_empty_strided op to become replicate from sharding
when necessary, this is a quick fix to resolve https://github.com/pytorch/pytorch/issues/107661

We'll need to think more about the behavior of this op when it comes to
sharding, one possibility is to follow the input sharding, but given the
output shape of this op might not be the same as the input, it's hard to
say we should follow the input sharding, further improvement needed once
we figure out the op syntax

wanchaol committed:
[dtensor] fix new_empty_strided op

This PR fixes the new_empty_strided op to become replicate from sharding
when necessary, this is a quick fix to resolve https://github.com/pytorch/pytorch/issues/107661

We'll need to think more about the behavior of this op when it comes to
sharding, one possibility is to follow the input sharding, but given the
output shape of this op might not be the same as the input, it's hard to
say we should follow the input sharding, further improvement needed once
we figure out the op syntax

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] fix new_empty_strided op"


This PR fixes the new_empty_strided op to become replicate from sharding
when necessary, this is a quick fix to resolve https://github.com/pytorch/pytorch/issues/107661

We'll need to think more about the behavior of this op when it comes to
sharding, one possibility is to follow the input sharding, but given the
output shape of this op might not be the same as the input, it's hard to
say we should follow the input sharding, further improvement needed once
we figure out the op syntax

[ghstack-poisoned]

fduwjj commented on test/distributed/_tensor/test_dtensor.py:
```diff
@@ -245,6 +245,22 @@ def test_to_local(self):
+        device_mesh = DeviceMesh(self.device_type, torch.arange(self.world_size))
+        local_tensor = torch.randn(8, 8, requires_grad=True, device=self.device_type)
+        my_dtensor = distribute_tensor(local_tensor, device_mesh, [Shard(0)])
+        new_strided_dtensor = my_dtensor.new_empty_strided((8, 8), (8, 1))
+        self.assertEqual(new_strided_dtensor.shape, my_dtensor.shape)
```
Any reason why we don't test BWD?

wanchaol replied:
you mean after new_empty_strided then sum().backward? I can write that test too

wanchaol committed:
Update on "[dtensor] fix new_empty_strided op"


This PR fixes the new_empty_strided op to become replicate from sharding
when necessary, this is a quick fix to resolve https://github.com/pytorch/pytorch/issues/107661

We'll need to think more about the behavior of this op when it comes to
sharding, one possibility is to follow the input sharding, but given the
output shape of this op might not be the same as the input, it's hard to
say we should follow the input sharding, further improvement needed once
we figure out the op syntax

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] fix new_empty_strided op"


This PR fixes the new_empty_strided op to become replicate from sharding
when necessary, this is a quick fix to resolve https://github.com/pytorch/pytorch/issues/107661

We'll need to think more about the behavior of this op when it comes to
sharding, one possibility is to follow the input sharding, but given the
output shape of this op might not be the same as the input, it's hard to
say we should follow the input sharding, further improvement needed once
we figure out the op syntax

[ghstack-poisoned]

fduwjj commented on test/distributed/_tensor/test_dtensor.py:
```diff
@@ -245,6 +245,28 @@ def test_to_local(self):
+        # test backward new_empty_strided with sharding works correctly
+        my_dtensor.to_local().sum().backward()
+        local_tensor.sum().backward()
+        self.assertEqual(
+            my_dtensor.grad.redistribute(placements=[Replicate()]).to_local(),
+            local_tensor.grad,
+        )
```
Do we also want to compare `new_strided_dtensor.grad` and `my_dtensor.grad` here?

wanchaol replied:
done :)

fduwjj reviewed (approved):
LGTM and one nit in the test.

wanchaol committed:
Update on "[dtensor] fix new_empty_strided op"


This PR fixes the new_empty_strided op to become replicate from sharding
when necessary, this is a quick fix to resolve https://github.com/pytorch/pytorch/issues/107661

We'll need to think more about the behavior of this op when it comes to
sharding, one possibility is to follow the input sharding, but given the
output shape of this op might not be the same as the input, it's hard to
say we should follow the input sharding, further improvement needed once
we figure out the op syntax

[ghstack-poisoned]

wanchaol commented:
@pytorchbot merge
